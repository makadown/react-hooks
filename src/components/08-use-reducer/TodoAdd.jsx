import React, { memo } from 'react';
import { useForm } from '../../hooks/useForm';

/**
 * Este es un fragmento que captura un texto y manda generar
 * un nuevo Todo
 */
export const TodoAdd = memo((props) => {
  const { handleAddTodo } = props;

  const [{ description }, handleInputChange, reset] = useForm({
    description: '',
  });

  const handleSubmit = (e) => {
    e.preventDefault();
    if (description.trim().length <= 1) {
      return;
    }
    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };
    // invoco la referencia de la funcion para que en el papa se 
    // agregue el todo
    handleAddTodo(newTodo);
    reset();
  };

  return (
    <>
      <h4> Agregar Todo </h4>
      <hr />
      <form onSubmit={handleSubmit}>
        <div className="d-grid gap-2">
          <input
            type="text"
            name="description"
            className="form-control"
            placeholder="Aprender..."
            autoComplete="off"
            value={description}
            onChange={handleInputChange}
          />
          <button type="submit" className="btn btn-outline-warning mt-1">
            Agregar
          </button>
        </div>
      </form>
    </>
  );
});

// Para que en los tests no se haga bolas con el memo
TodoAdd.displayName='TodoAdd';