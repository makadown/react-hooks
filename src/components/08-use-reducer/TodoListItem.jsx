import React, { memo } from 'react';

export const TodoListItem = memo((props) => {
  // console.log('props de todolistitem', props);
  const { todo, index, handleDelete, handleToggle } = props;

  let nombreClase = 'none';
  
  if (todo.done && todo.done === true) {
    nombreClase = 'complete disabled';
  }

  return (
    <li
      key={todo.id}
      className="list-group-item list-group-item-dark d-flex justify-content-between align-items-start"
    >
      <div className="ms-2 me-auto">
        <div className="fw-bold">
          <p
            className={nombreClase}
            onClick={() => handleToggle(todo.id)}
          >
            {index + 1}.- {todo.desc}
          </p>
        </div>
        { todo.done ? 'Finalizado' : 'Pendiente'}
      </div>
      <span>
        <button disabled={!todo.done}
          className="btn btn-danger"
          onClick={() => handleDelete(todo.id)}
        >
          Borrar
        </button>
      </span>
    </li>
  );
});
