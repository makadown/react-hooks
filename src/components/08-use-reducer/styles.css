body {
    padding: 50px;
}

p {
    cursor: pointer;
    margin: 0px;
}

.list-group-flush {
    border-radius: 5px;
}

.list-group-item {
    border-radius: 4px;
    align-items: center;
    display:flex;
    justify-content: space-between;
}

.complete {
    text-decoration: line-through;
}