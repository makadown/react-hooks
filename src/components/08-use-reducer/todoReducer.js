
export const todoReducer = (state = [], action) => {
    switch (action.type) {
       case 'add':
           // aqui el payload es un objeto TODO completo
           return [...state, action.payload];
        case 'delete':
            // aqui el payload es el id
            return state.filter(todo => todo.id !== action.payload );
        case 'toggle':
            // aqui el payload es el id
            return state.map( todo => 
                ( todo.id === action.payload ) ? 
                    { ...todo, done: !todo.done } :
                    todo
            );
        case 'toggle-legacy':
                return state.map( todo => {
                    if ( todo.id === action.payload ) {
                        return {
                            ...todo,
                            done: !todo.done
                        };
                    } else { return todo; }
                });
        default:
            return state;
    }
}