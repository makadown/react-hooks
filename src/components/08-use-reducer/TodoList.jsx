import React, { memo } from 'react'
import { TodoListItem } from './TodoListItem';

export const TodoList = memo((props) => {
    const { todos, handleDelete, handleToggle } = props;

    return (
        <ul className="list-group list-group-flush">
                        {
                            todos.map((todo, index) => (
                                <TodoListItem todo={todo} 
                                        key={todo.id}
                                        index={index} 
                                        handleDelete={handleDelete} 
                                        handleToggle={handleToggle}>

                                </TodoListItem>
                            ))
                        }
        </ul>
    )
}
);
// Para que en los tests no se haga bolas con el memo
TodoList.displayName='TodoList';