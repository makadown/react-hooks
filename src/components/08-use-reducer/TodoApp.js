import React, { memo, useEffect, useReducer } from 'react';
import './styles.css';
import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

const init = () => {
    // Si no hay localstorage (null), regresa un arreglo vacío
    return JSON.parse(localStorage.getItem('todos')) || [];
};

export const TodoApp = memo(() => {
    const [todos, dispatch] = useReducer(todoReducer, [], init);

    // Este useEffect me ayudará a almacenar en localstore cada que se 
    // detecte un cambio en el arreglo de todos!
    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos])

    const handleDelete = (todoId) => {
        console.log(todoId);
        // crear la action para borrar
        const action = {
            type: 'delete',
            payload: todoId
        };
        // dispatch
        dispatch(action);
    };

    const handleToggle = (todoId) => {
        console.log(todoId);
        // crear la action para borrar
        const action = {
            type: 'toggle',
            payload: todoId
        };
        // dispatch
        dispatch(action);
    };

    const handleAddTodo = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        });
    };

    return (
        <div>
            <h1>Todo App ( {todos.length} )</h1>
            <hr />
            <div className="row">
                <div className="col-7">
                    {/* Crear componente TodoList requiriendo 
                        todos, handleDelete, handleToggle
                    */}
                    <TodoList todos={todos}
                        handleDelete={handleDelete}
                        handleToggle={handleToggle}>
                    </TodoList>
                </div>
                <div className="col-5">
                    <TodoAdd handleAddTodo={handleAddTodo}></TodoAdd>
                    <hr />
                    {
                        todos.length > 0 &&
                        <small>Haz click en cada tarea para darla como hecha.</small>
                    }
                </div>
            </div>
        </div>
    );
});
