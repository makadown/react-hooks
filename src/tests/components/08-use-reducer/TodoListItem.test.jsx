import {shallow} from 'enzyme';
import { TodoListItem } from '../../../components/08-use-reducer/TodoListItem';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Pruebas en componente TodoListItem', () => {
    
    const singleTodo = demoTodos[0];
    const index = 0;
    let handleDelete = jest.fn();
    let handleToggle = jest.fn();
    
    let wrapper = shallow(<TodoListItem todo={singleTodo} index={index}
                    handleDelete={handleDelete} handleToggle={handleToggle}>
                        </TodoListItem>);

    beforeEach(() => {
        wrapper = shallow(<TodoListItem todo={singleTodo} index={index}
            handleDelete={handleDelete} handleToggle={handleToggle}>
                </TodoListItem>);
    });

    test('Debe de mostrar TodoListItem correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('TodoListItem debe llamar la funcion handleDelete', () => {
        wrapper.find('button').simulate('click');
        // expect(handleDelete).toHaveBeenCalled();
        expect(handleDelete).toHaveBeenCalledWith(singleTodo.id);
    });
   
    test('TodoListItem debe llamar la funcion handleToggle', () => {
        wrapper.find('p').simulate('click');
        // expect(handleToggle).toHaveBeenCalled();
        expect(handleToggle).toHaveBeenCalledWith(singleTodo.id);
    });
     
    test('TodoListItem debe mostrar el texto correctamente', () => {
        const contenido = wrapper.find('p').text();
        expect(contenido.trim().length).toBeGreaterThan(5);
        expect(contenido).toContain('1.-');
        expect(contenido).toContain(singleTodo.desc);
    });

    test('Debe tener la clase complete si el todo es done!', () => {
        let singleTodoDone = {...singleTodo};
        singleTodoDone.done = true;
        const otroWrapper = 
                shallow(<TodoListItem todo={singleTodoDone}></TodoListItem>);
        expect(otroWrapper.find('.complete').exists()).toBeTruthy();        
    });
});
