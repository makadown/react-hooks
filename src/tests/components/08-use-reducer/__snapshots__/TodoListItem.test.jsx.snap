// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente TodoListItem Debe de mostrar TodoListItem correctamente 1`] = `
<li
  className="list-group-item list-group-item-dark d-flex justify-content-between align-items-start"
  key="1"
>
  <div
    className="ms-2 me-auto"
  >
    <div
      className="fw-bold"
    >
      <p
        className="none"
        onClick={[Function]}
      >
        1
        .- 
        Aprender React
      </p>
    </div>
    Pendiente
  </div>
  <span>
    <button
      className="btn btn-danger"
      disabled={true}
      onClick={[Function]}
    >
      Borrar
    </button>
  </span>
</li>
`;
